azure_monitor/src/azure_monitor/protocol.py

```diff
@@ -2,188 +2,347 @@
-        return repr(tmp)
-
-    def __setattr__(self, name, value):
-        self[name] = value
-
-    def __getattr__(self, name):
-        try:
-            return self[name]
-        except KeyError:
-            raise AttributeError(
-                "'{}' object has no attribute {}".format(
-                    type(self).__name__, name
-                )
-            )
-
-    def __getitem__(self, key):
-        if self._default is self:
-            return super().__getitem__(key)
-        if key in self:
-            return super().__getitem__(key)
-        return self._default[key]
 
+        for key in self.__slots__:
+            data = getattr(self, key, None)
+            if isinstance(data, BaseObject):
+                tmp[key] = repr(data)
+            else:
+                tmp[key] = data
 
-BaseObject._default = BaseObject()  # noqa pylint: disable=protected-access
+        return repr(tmp)
 
 
 class Data(BaseObject):
-    _default = BaseObject(baseData=None, baseType=None)
+    __slots__ = ("base_data", "base_type")
+
+    def __init__(self, base_data=None, base_type=None) -> None:
+        self.base_data = base_data
+        self.base_type = base_type
 
-    def __init__(self, *args, **kwargs):
-        super().__init__(*args, **kwargs)
-        self.baseData = self.baseData  # noqa pylint: disable=invalid-name
-        self.baseType = self.baseType  # noqa pylint: disable=invalid-name
+    def to_dict(self):
+        return {
+            "baseData": self.base_data.to_dict(),
+            "baseType": self.base_type,
+        }
 
 
 class DataPoint(BaseObject):
-    _default = BaseObject(
+    __slots__ = (
+        "ns",
+        "name",
+        "kind",
+        "value",
+        "count",
+        "min",
+        "max",
+        "std_dev",
+    )
+
+    def __init__(
+        self,
         ns="",
         name="",
         kind=None,
         value=0.0,
         count=None,
         min=None,
         max=None,
-        stdDev=None,
-    )
-
-    def __init__(self, *args, **kwargs):
-        super().__init__(*args, **kwargs)
-        self.name = self.name
-        self.value = self.value
+        std_dev=None,
+    ) -> None:
+        self.ns = ns
+        self.name = name
+        self.kind = kind
+        self.value = value
+        self.count = count
+        self.min = min
+        self.max = max
+        self.std_dev = std_dev
+
+    def to_dict(self):
+        return {
+            "ns": self.ns,
+            "name": self.name,
+            "kind": self.kind,
+            "value": self.value,
+            "count": self.count,
+            "min": self.min,
+            "max": self.max,
+            "stdDev": self.std_dev,
+        }
 
 
 class Envelope(BaseObject):
-    _default = BaseObject(
+    __slots__ = (
+        "ver",
+        "name",
+        "time",
+        "sample_rate",
+        "seq",
+        "ikey",
+        "flags",
+        "tags",
+        "data",
+        "base_type"
```

Comment:
I don't think envelope needs `base_type`

Reply:
I added it because it was being used on `tests/test_base_exporter.py`!

Reply:
If this is not going to be used, I can change the test to use the `base_type` inside `Data` and add it to the envelope.

Reply:
It is needed, if is not in there people would not be able to modify the value in telemetry processors

Reply:
One question, why edit the `base_type`? Is this the value that defines, for example, if it is an `Azure Blob Storage`, `SQL`, `HTTP Request` and etc for the Application Insights?

Reply:
Ok, we double checked this and baseType is inside data property so we need to update the tests, updating baseType makes sense in customEvents in AppInsights

Reply:
Ok, will update them here

Reply:
Done!

Reply:
> One question, why edit the `base_type`? Is this the value that defines, for example, if it is an `Azure Blob Storage`, `SQL`, `HTTP Request` and etc for the Application Insights?

Application Insights uses "baseType" to identify what kind of telemetry is being sent (`requests`, `dependencies`, `traces`, `exception`, etc..).

Reply:
In case of the `Data` `base_type` yes! I was wondering about the `Envelope` one, if it was related to something else